File: medium/cpp/1382.cpp
// https://leetcode.com/problems/balance-a-binary-search-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    TreeNode* balanceBST(TreeNode* root)
    {
        std::vector<int> sorted;
        fill_sorted(root, sorted);
        
        return build(sorted, 0, std::size(sorted));
    }
    
private:
    void fill_sorted(TreeNode* node, std::vector<int>& sorted)
    {
        if (node == nullptr)
        {
            return;
        }
        
        fill_sorted(node->left, sorted);
        sorted.push_back(node->val);
        fill_sorted(node->right, sorted);
    }
    
    TreeNode* build(const std::vector<int>& sorted, std::size_t left, std::size_t right)
    {
        if (left >= right)
        {
            return nullptr;
        }
        
        const std::size_t mid = left + (right - left) / 2;
        
        auto node = new TreeNode(sorted[mid]);
        
        node->left = build(sorted, left, mid);
        node->right = build(sorted, mid + 1, right);
        
        return node;
    }
};

File: medium/python/1382.py
# https://leetcode.com/problems/balance-a-binary-search-tree/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        values = []
        self.traverse(root, values)
        return self.build(values, 0, len(values))

    def traverse(self, root, values):
        if root is not None:
            self.traverse(root.left, values)
            values.append(root.val)
            self.traverse(root.right, values)

    def build(self, values, left, right):
        if left < right:
            mid = left + (right - left) // 2
            return TreeNode(values[mid], self.build(values, left, mid), self.build(values, mid + 1, right))